### PR TITLE
Auto-embed trace information into TypedHttpClient request headers

### PR DESCRIPTION
**Pull request by nbroyles**
Additionally, make TracingInterceptor aware of potential parent
span information in headers and use it to set parent context, if it's
available.

### CODE REVIEW

File: misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  private val multimap: Map<String, List<String>>
+) : TextMap {
+  override fun put(key: String?, value: String?) {
+    throw UnsupportedOperationException("Cannot put into readonly data source")
+  }
+
+  override fun iterator(): MutableIterator<MutableMap.MutableEntry<String, String>> {
+    return multimap.map {
+      val values = it.value
+      val mapEntries = values.map { mapValue -> ExtractEntry(it.key, mapValue) }
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt]: nit, don't need the variable declaration

File: misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+class TextMultimapExtractAdapter(
+  private val multimap: Map<String, List<String>>
+) : TextMap {
+  override fun put(key: String?, value: String?) {
+    throw UnsupportedOperationException("Cannot put into readonly data source")
+  }
+
+  override fun iterator(): MutableIterator<MutableMap.MutableEntry<String, String>> {
+    return multimap.map {
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt]: nit: use `flatMap` rather than `flatten`

File: misk/src/test/kotlin/misk/client/TypedHttpClientTest.kt [hunk 3]
```diff
@@ -42,22 +48,64 @@ internal class TypedHttpClientTest {
   @Inject
   private lateinit var jetty: JettyService
 
-  private lateinit var client: ReturnADinosaur
+  @Inject
+  private lateinit var tracer: Tracer
```
Comment (mmihic) [inline misk/src/test/kotlin/misk/client/TypedHttpClientTest.kt]: Let's keep test cases focused - move trace testing to another dedicate test.

File: misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt [hunk 4]
```diff
@@ -29,7 +32,21 @@ internal class TracingInterceptor internal constructor(private val tracer: Trace
   }
 
   override fun intercept(chain: NetworkChain): Response<*> {
-    val scope = tracer.buildSpan("${chain.action}").startActive(true)
+    var parentContext: SpanContext? = null
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt]: can make this a val by doing:

````kotlin
val parentContext = try {
   tracer.extract(Format.Builtin.HTTP_HEADERS, TextMultimapExtractAdapter(chain.request.headers.toMultimap()))
} catch (e: Exception) {
   logger.info("blah blah")
   null
}

````

File: misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt [hunk 5]
```diff
@@ -29,7 +32,21 @@ internal class TracingInterceptor internal constructor(private val tracer: Trace
...
+    try {
+      parentContext = tracer.extract(Format.Builtin.HTTP_HEADERS,
+          TextMultimapExtractAdapter(chain.request.headers.toMultimap()))
+    } catch (exception: Exception) {
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt]: nit: Use ``e`` for exception name

File: misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt [hunk 6]
```diff
@@ -29,7 +32,21 @@ internal class TracingInterceptor internal constructor(private val tracer: Trace
...
+      logger.info("Failure attempting to extract span context. Existing context, if any," +
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt]: What can cause this failure? Should this be at info level logging?

File: misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import io.opentracing.propagation.TextMap
+
+/**
+ * Provides an interface for Tracer.extract() to read from a multimap datasource
+ * (i.e. HTTP headers provided via OkHttp)
+ */
+class TextMultimapExtractAdapter(
+  private val multimap: Map<String, List<String>>
+) : TextMap {
+  override fun put(key: String?, value: String?) {
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt]: nit: these are technically not nullable, are they?

File: misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+  override fun iterator(): MutableIterator<MutableMap.MutableEntry<String, String>> {
+    return multimap.map {
+      val values = it.value
+      val mapEntries = values.map { mapValue -> ExtractEntry(it.key, mapValue) }
+      mapEntries
+    }.flatten().toMutableList().iterator()
+  }
+
+  internal class ExtractEntry(
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt]: Should this be ``internal``, or ``private``?

File: misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package misk.tracing.interceptors
+
+import io.opentracing.propagation.TextMap
+
+/**
+ * Provides an interface for Tracer.extract() to read from a multimap datasource
+ * (i.e. HTTP headers provided via OkHttp)
+ */
+class TextMultimapExtractAdapter(
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt]: Does this need to be ``public`` or can it be ``internal``?

File: misk/src/test/kotlin/misk/client/TypedHttpClientTest.kt [hunk 10]
```diff
@@ -42,22 +48,64 @@ internal class TypedHttpClientTest {
...
-    val response = client.getDinosaur(Dinosaur.Builder().name("trex").build()).execute()
+    val client = clientInjector.getInstance(ReturnADinosaur::class.java)
+    val response = client.getDinosaur(dinosaurRequest).execute()
     assertThat(response.code()).isEqualTo(200)
     assertThat(response.body()).isNotNull()
     assertThat(response.body()?.name!!).isEqualTo("supertrex")
   }
 
+  @Test
+  fun embedsTrace() {
```
Comment (mmihic) [inline misk/src/test/kotlin/misk/client/TypedHttpClientTest.kt]: We should also have a test that does multiple hops (from client to server to server again), to ensure that parent span ids propagate.

File: misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt (hunk 6 above)
Comment (nbroyles) [reply misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt]: Corrupted trace data (e.g. mangled header) can cause a failure. Thinking about it, we probably want to bump this to warn. Should be rare enough and people probably want to know.

File: misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt (hunk 7 above)
Comment (nbroyles) [reply misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt]: Good catch. Not sure what happened here

File: misk/src/main/kotlin/misk/client/ClientInvocationHandler.kt [hunk 11]
```diff
@@ -56,8 +59,12 @@ internal class ClientInvocationHandler(
           clientBuilder.build()
         }
 
-    methodName to retrofit.newBuilder()
+    val retrofitBuilder = retrofit.newBuilder()
         .client(actionSpecificClient)
+
+    if (tracer != null) retrofitBuilder.callFactory(TracingCallFactory(actionSpecificClient, tracer))
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/client/ClientInvocationHandler.kt]: Please add this to the retrofit object created in the `TypedHttpClientModule` rather than doing it here. All global retrofit properties should be established in the module; anything added here is specific to an individual action

File: misk/src/main/kotlin/misk/client/TypedHttpClientModule.kt [hunk 12]
```diff
@@ -61,21 +62,24 @@ class TypedHttpClientModule<T : Any>(
...
+    @Inject(optional=true)
+    private val tracer: Tracer? = null
+
     override fun get(): T {
       val okhttp = httpClientProvider.get()
 
       val retrofit = Retrofit.Builder()
           .baseUrl(httpClientsConfig[name].url)
           .addConverterFactory(MoshiConverterFactory.create(moshi))
-          .build()
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/client/TypedHttpClientModule.kt]: Why this change?

File: misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+internal class TextMultimapExtractAdapter(
+  private val multimap: Map<String, List<String>>
+) : TextMap {
+  override fun put(key: String, value: String) {
+    throw UnsupportedOperationException("Cannot put into readonly data source")
+  }
+
+  override fun iterator(): MutableIterator<MutableMap.MutableEntry<String, String>> {
+    return multimap.flatMap {
+      val values = it.value
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/tracing/interceptors/TextMultimapExtractAdapter.kt]: nit. just ``it.values.map { mapValues -> ExtractEntry(it.key, mapValue) }``

File: misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt [hunk 14]
```diff
@@ -29,7 +32,21 @@ internal class TracingInterceptor internal constructor(private val tracer: Trace
   }
 
   override fun intercept(chain: NetworkChain): Response<*> {
-    val scope = tracer.buildSpan("${chain.action}").startActive(true)
+    val parentContext: SpanContext? = try {
+      tracer.extract(Format.Builtin.HTTP_HEADERS,
+          TextMultimapExtractAdapter(chain.request.headers.toMultimap()))
+    } catch (exception: Exception) {
```
Comment (mmihic) [inline misk/src/main/kotlin/misk/web/interceptors/TracingInterceptor.kt]: nit: use ``e`` not ``exception``

File: misk/src/main/kotlin/misk/client/TypedHttpClientModule.kt (hunk 12 above)
Comment (nbroyles) [reply misk/src/main/kotlin/misk/client/TypedHttpClientModule.kt]: Blerg, merge conflict resolution gone bad. Will fix.

File: misk/src/main/kotlin/misk/client/ClientInvocationHandler.kt (hunk 11 above)
Comment (nbroyles) [reply misk/src/main/kotlin/misk/client/ClientInvocationHandler.kt]: So, I originally had it in the `TypedHttpClientModule`, but that broke once the `ClientInvocationHandler` changes landed. The issue was that the `TracingCallFactory` wasn't  being given the correct `OkHttpClient` after the changes. The `OkHttpClient` in `TypedHttpClientModule` isn't the client performing the actual calls, rather a template to generate the `actionSpecificClient`. Therefore, to get tracing to work properly, it needs to be installed in the `ClientInvocationHandler` the the call factory can be applied to the correct `OkHttpClient`
